src/appleseed/renderer/kernel/lighting/lighttypes.cpp

```diff
@@ -0,0 +1,142 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "lighttypes.h"
+
+// appleseed. renderer headers.
+#include "renderer/modeling/input/source.h"
+#include "renderer/modeling/light/light.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/transform.h"
+
+namespace renderer
+{
+
+LightSource::LightSource()
+{
+}
+
+LightSource::~LightSource()
+{
+}
+
+//
+// Non-physical light source class implementation.
+//
+
+NonPhysicalLightSource::NonPhysicalLightSource(
+    const NonPhysicalLightInfo* light)
+  : m_light_info(light)
+{
+}
+
+foundation::Vector3d NonPhysicalLightSource::get_position()  const
+{
+    const Light* light = m_light_info->m_light;
+
+    // Compute the exact position of the light.
+    foundation::Vector3d position = light->get_transform()
+                                          .get_local_to_parent()
+                                          .extract_translation();
+    
+    return position;
+}
+
+foundation::AABB3d NonPhysicalLightSource::get_bbox() const
+{
+    const Light* light = m_light_info->m_light;
+    foundation::Vector3d position = light->get_transform()
+                                          .get_local_to_parent()
+                                          .extract_translation();
+
+    // Non physical light has no real size - hence we are assigning it some
+    // arbitrary small value for the bbox.
+    return foundation::AABB3d(
+                foundation::Vector3d(position[0] - 0.001,
+                                     position[1] - 0.001,
+                                     position[2] - 0.001),
+                foundation::Vector3d(position[0] + 0.001,
+                                     position[1] + 0.001,
+                                     position[2] + 0.001));
+}
+
+Spectrum NonPhysicalLightSource::get_intensity() const
+{
+    const Light* light = m_light_info->m_light;
+    Spectrum intensity;
+    light->get_inputs().find("intensity").source()->evaluate_uniform(intensity);
+    return intensity;
+}
+
+//
+// Emitting triangle light source class implementation.
+//
+
+EmittingTriangleLightSource::EmittingTriangleLightSource(
+    const EmittingTriangle* light)
+  : m_light(light)
+{
+}
+
+foundation::Vector3d EmittingTriangleLightSource::get_position() const
+{
+    // Retrieve the coordinates of each vertex in world space.
```

Comment:
What about getting rid of the intermediary variables? It will be easier to read and potentially more efficient:

    foundation::Vector3d EmittingTriangleLightSource::get_position() const
    {
        return (m_light->m_v0 + m_light->m_v1 + m_light->m_v2) / 3;
    }

You don't even need the comment that doesn't add much value.